tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -0,0 +1,267 @@
+        return ModeMonitor(
+            center=port.center,
+            size=port.size,
+            freqs=[self.freq],
+            mode_spec=port.mode_spec,
+            name=port.name,
+        )
+
+    def _to_sources(self, port: Port) -> List[ModeSource]:
+        """Creates a list of mode sources from a given port."""
+        return [
+            ModeSource(
+                center=port.center,
+                size=port.size,
+                source_time=GaussianPulse(freq0=self.freq, fwidth=self.freq * FWIDTH_FRAC),
+                mode_spec=port.mode_spec,
+                mode_index=mode_index,
+                direction=port.direction,
+                name=port.name,
+            )
+            for mode_index in port.mode_indices
+        ]
+
+    @equal_aspect
+    @add_ax_if_none
+    def plot_sim(self, x: float = None, y: float = None, z: float = None, ax: Ax = None) -> Ax:
+        """Plot a :class:`Simulation` with all sources added for each port, for troubleshooting."""
+
+        sim_plot = self.simulation.copy(deep=True)
+        for port_source in self.ports:
+            mode_source_0 = self._to_sources(port_source)[0]
+            sim_plot.sources.append(mode_source_0)
+        return sim_plot.plot(x=x, y=y, z=z, ax=ax)
+
+    def _shift_value(self, port: Port) -> float:
+        """How far (signed) to shift the monitor from the source."""
+        normal_index = port.size.index(0.0)
+        dl = self.simulation.grid_size[normal_index]
+        if not isinstance(dl, float):
+            raise NotImplementedError("doesn't support nonuniform. How many grid cells to shift?")
+        return dl if port.direction == "+" else -1 * dl
+
+    def _shift_port(self, port: Port) -> Port:
+        """Generate a new port shifted by one grid cell in normal direction."""
+
+        shift_value = self._shift_value(port)
+        center_shifted = list(port.center)
+        center_shifted[port.size.index(0.0)] += shift_value
+        port_shifted = port.copy(deep=True)
+        port_shifted.center = center_shifted
+        return port_shifted
+
+    def _task_name(self, port_source: Port, mode_index: int) -> str:
+        """The name of a task, determined by the port of the source and mode index."""
+        return f"smatrix_port{port_source.name}_mode{mode_index}"
+
+    def _make_sims(self) -> Dict[str, Simulation]:
+        """Generate all the :class:`Simulation` objects for the S matrix calculation."""
+        sim_dict = {}
+        for port_source in self.ports:
+            for mode_source in self._to_sources(port_source):
+                sim_copy = self.simulation.copy(deep=True)
+                sim_copy.sources = [mode_source]
+                for port_monitor in self.ports:
+                    if port_source == port_monitor:
+                        port_monitor = self._shift_port(port_source)
+                    mode_monitor = self._to_monitor(port_monitor)
+                    sim_copy.monitors.append(mode_monitor)
```

Comment:
Just to clarify, the user is allowed to have specified other monitors beforehand, unrelated to the S matrix computation? I guess that's fine since they're not interfering with the simulation, but if there's data download involved, then maybe a `_sim_has_no_monitors` validator could be introduced?

Reply:
After having seen the notebook though, looks like it's actually quite useful to allow specifying additional monitors.

Reply:
yea originally I had forbidden extra monitors but after playing with the tool I put them back in because it was very useful.  If the user leaves monitors in their original simulation, I assume they know what they are doing, but maybe eventually we could add an `exclude_extra_monitors` flag to the `ComponentModeler` to copy over `simulation` initially and remove all monitors for the purpose of computing S matrix.